### PR TITLE
✨ clusterctl: migrate CRDs during clusterctl upgrade

### LINKED ISSUE #6674: Clusterctl upgrade fails due to cert-manager CRDs
**What steps did you take and what happened:**
When trying to upgrade our clusters to the latest capi version ( 1.1.4 ) we noticed that the upgrade fails
```
clusterctl upgrade apply --contract v1beta1
Checking cert-manager version...
Deleting cert-manager Version="v1.5.3"
Installing cert-manager Version="v1.7.2"
Error: action failed after 10 attempts: failed to update cert-manager component apiextensions.k8s.io/v1, Kind=CustomResourceDefinition, /certificaterequests.cert-manager.io: CustomResourceDefinition.apiextensions.k8s.io "certificaterequests.cert-manager.io" is invalid: status.storedVersions[0]: Invalid value: "v1alpha2": must appear in spec.versions
```
After closer inspection that seems to be an issue caused by https://github.com/kubernetes-sigs/cluster-api/pull/6432 , as v1.6 got skipped.
1.6 marked alpha2 as not served, so only after applying 1.6 crds the upgrade to 1.7 is possible
https://github.com/cert-manager/cert-manager/commit/c6896b2f93e7d2f9db71dd5ed6c637acfaf31a0b

**What did you expect to happen:**
Clusterctl correctly upgrades cert-manager

**Environment:**

- Cluster-api version:
- Minikube/KIND version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

/kind bug

### LINKED ISSUE #6691: Handle status.storedVersions migration during upgrades
**User Story**

As a developer/user/operator, I would like to do an upgrade from old API versions to newer ones in sequence (v1a3 => v1a4 => v1beta1 ) without manually removing stored versions and applying them in the middle of the upgrade process (v1a4 => v1beta1 since v1a4 would have v1a3 `storedVersion` in the status). 

**Detailed Description**

we have a scenario where we do an upgrade from CAPM3 v1a4 => v1a5 => v1beta1 (CAPI  v1a3 => v1a4 => v1beta1). The problem is, clusterctl fails (see logs below) to upgrade from v1a5 => v1b1 because of the `status.storedVersions` in the CRD which includes `v1a4` API version in the status. 

```
Patching CustomResourceDefinition="metal3clusters.infrastructure.cluster.x-k8s.io"
Retrying with backoff Cause="failed to patch provider object: CustomResourceDefinition.apiextensions.k8s.io \"metal3clusters.infrastructure.cluster.x-k8s.io\" is invalid: status.storedVersions[0]: Invalid value: \"v1alpha4\": must appear in spec.versions"
Patching CustomResourceDefinition="metal3clusters.infrastructure.cluster.x-k8s.io"
Retrying with backoff Cause="failed to patch provider object: CustomResourceDefinition.apiextensions.k8s.io \"metal3clusters.infrastructure.cluster.x-k8s.io\" is invalid: status.storedVersions[0]: Invalid value: \"v1alpha4\": must appear in spec.versions"
Patching CustomResourceDefinition="metal3clusters.infrastructure.cluster.x-k8s.io"
Retrying with backoff Cause="failed to patch provider object: CustomResourceDefinition.apiextensions.k8s.io \"metal3clusters.infrastructure.cluster.x-k8s.io\" is invalid: status.storedVersions[0]: Invalid value: \"v1alpha4\": must appear in spec.versions"
Patching CustomResourceDefinition="metal3clusters.infrastructure.cluster.x-k8s.io"
Retrying with backoff Cause="failed to patch provider object: CustomResourceDefinition.apiextensions.k8s.io \"metal3clusters.infrastructure.cluster.x-k8s.io\" is invalid: status.storedVersions[0]: Invalid value: \"v1alpha4\": must appear in spec.versions"
Patching CustomResourceDefinition="metal3clusters.infrastructure.cluster.x-k8s.io"
Retrying with backoff Cause="failed to patch provider object: CustomResourceDefinition.apiextensions.k8s.io \"metal3clusters.infrastructure.cluster.x-k8s.io\" is invalid: status.storedVersions[0]: Invalid value: \"v1alpha4\": must appear in spec.versions"
Patching CustomResourceDefinition="metal3clusters.infrastructure.cluster.x-k8s.io"
Retrying with backoff Cause="failed to patch provider object: CustomResourceDefinition.apiextensions.k8s.io \"metal3clusters.infrastructure.cluster.x-k8s.io\" is invalid: status.storedVersions[0]: Invalid value: \"v1alpha4\": must appear in spec.versions"
Patching CustomResourceDefinition="metal3clusters.infrastructure.cluster.x-k8s.io"
Retrying with backoff Cause="failed to patch provider object: CustomResourceDefinition.apiextensions.k8s.io \"metal3clusters.infrastructure.cluster.x-k8s.io\" is invalid: status.storedVersions[0]: Invalid value: \"v1alpha4\": must appear in spec.versions"
Patching CustomResourceDefinition="metal3clusters.infrastructure.cluster.x-k8s.io"
Retrying with backoff Cause="failed to patch provider object: CustomResourceDefinition.apiextensions.k8s.io \"metal3clusters.infrastructure.cluster.x-k8s.io\" is invalid: status.storedVersions[0]: Invalid value: \"v1alpha4\": must appear in spec.versions"
Patching CustomResourceDefinition="metal3clusters.infrastructure.cluster.x-k8s.io"
Retrying with backoff Cause="failed to patch provider object: CustomResourceDefinition.apiextensions.k8s.io \"metal3clusters.infrastructure.cluster.x-k8s.io\" is invalid: status.storedVersions[0]: Invalid value: \"v1alpha4\": must appear in spec.versions"
Patching CustomResourceDefinition="metal3clusters.infrastructure.cluster.x-k8s.io"
```

 The request would be, should the clusterctl take care of removing of old version from the `status.storedVersions` or how is it handled in general?
 

**Anything else you would like to add:**

There has been a long discussion in the slack thread, xref: https://kubernetes.slack.com/archives/C8TSNPY4T/p1655805341632159

/kind feature

### PR DESCRIPTION
**Pull request by sbueringer**
Credits to @fabriziopandini for the initial implementation

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #6674
Fixes #6691

### CODE REVIEW

**Comment by sbueringer** [comment]
PR should be ready for review. I wrote something about the limitations here: https://github.com/kubernetes-sigs/cluster-api/issues/6674#issuecomment-1167467696

**Comment by fabriziopandini** [comment]
@furkatgofurov7 @schrej it will be great if you can test if the proposed solution fixes the issue

File: cmd/clusterctl/client/cluster/crd_migration.go [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+		}); err != nil {
+			return errors.Wrapf(err, "failed to list %q", list.GetKind())
+		}
+
+		for i := range list.Items {
+			obj := list.Items[i]
+
+			log.V(5).Info("Migrating", logf.UnstructuredToValues(obj)...)
+			if err := retryWithExponentialBackoff(newWriteBackoff(), func() error {
+				return handleMigrateErr(m.Client.Update(ctx, &obj))
```
Comment (sbueringer) [inline cmd/clusterctl/client/cluster/crd_migration.go:163]: Note to myself: verify how/if that changes managed fields of topology owned resources

File: cmd/clusterctl/client/cluster/crd_migration.go [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+			if _, err := m.run(ctx, crd); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Run migrates CRs of a new CRD.
```
Comment (chrischdi) [inline cmd/clusterctl/client/cluster/crd_migration.go:73]:
```suggestion
// run migrates CRs of a new CRD.
```

**Review by chrischdi — COMMENTED** [review]
Code looks very clean and good to me.

Just some small nits / questions from my side 👍

File: cmd/clusterctl/client/cluster/crd_migration.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+	// If the new CRD still contains all current stored versions, nothing to do
+	// as no previous storage version will be dropped.
+	if newVersions.HasAll(currentStatusStoredVersions.List()...) {
+		log.V(2).Info("CRD migration check passed", "name", newCRD.Name)
+		return false, nil
+	}
+
+	// Otherwise a version that has been used as storage version will be dropped, so it is necessary to migrate all the
+	// objects and drop the storage version from the current CRD status before installing the new CRD.
+	// Rif https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#writing-reading-and-updating-versioned-customresourcedefinition-objects
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply cmd/clusterctl/client/cluster/crd_migration.go:119]: Couldn't find a good definition that matched except Reading is Fundamental :) 
I'll replace it

File: cmd/clusterctl/client/cluster/crd_migration.go [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+	storedVersionsToDelete := currentStatusStoredVersions.Difference(newVersions)
+	storedVersionsToPreserve := currentStatusStoredVersions.Intersection(newVersions)
+	log.Info("CR migration required", "kind", newCRD.Spec.Names.Kind, "storedVersionsToDelete", strings.Join(storedVersionsToDelete.List(), ","), "storedVersionsToPreserve", strings.Join(storedVersionsToPreserve.List(), ","))
+
+	if err := m.migrateResourcesForCRD(ctx, currentCRD, currentStorageVersion); err != nil {
+		return false, err
+	}
+
+	if err := m.patchCRDStoredVersions(ctx, currentCRD, storedVersionsToPreserve.List()); err != nil {
+		return false, err
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply cmd/clusterctl/client/cluster/crd_migration.go:133]: Currently it's the intersection of newVersions and currentStatusStoredVersions not all newVersions.

But yeah after the update they should all have the current storage version so we can avoid additional migrations in later upgrades by just setting the current storage version.

Good finding

File: cmd/clusterctl/main.go [hunk 5]
```diff
@@ -17,11 +17,16 @@ limitations under the License.
...
+	"math/rand"
+	"time"
+
 	_ "k8s.io/client-go/plugin/pkg/client/auth"
 
 	"sigs.k8s.io/cluster-api/cmd/clusterctl/cmd"
 )
 
 func main() {
+	rand.Seed(time.Now().UnixNano())
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply cmd/clusterctl/main.go:29]: In general yes, we don't have that comment anywhere else so I can only guess why it was added.

Maybe for exponential backoff to be truly random? But if that's the case who important is that that for every clusterctl run we have a new random sequence?

Maybe I should rather drop it

**Comment by MaxRink** [comment]
@fabriziopandini @sbueringer  we just tried that upgrading with this PR and it failed upgrading metal³
logs: https://gist.github.com/MaxRink/fa3b27b4193a1a4bb484e47197713b88

**Comment by sbueringer** [comment]
> @fabriziopandini @sbueringer we just tried that upgrading with this PR and it failed upgrading metal³ logs: https://gist.github.com/MaxRink/fa3b27b4193a1a4bb484e47197713b88 On a first glance it looks to me linke the order of operation is wrong, the migration should run before any of the providers get scaled down so that conversion webhooks are still available

Thx that sounds absolutely correct.. I'll update the PR

File: cmd/clusterctl/client/cluster/crd_migration.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sbueringer) [reply cmd/clusterctl/client/cluster/crd_migration.go:133]: Done

**Comment by sbueringer** [comment]
@MaxRink I don't have time right now to test it again manually, but if you want you can give it another try.

I'll probably do more testing tomorrow.

/test pull-cluster-api-e2e-full-main

File: cmd/clusterctl/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (sbueringer) [reply cmd/clusterctl/main.go:29]: Looks like the apimachinery rand package uses it's own random source: [mrand ](https://github.com/kubernetes/apimachinery/blob/master/pkg/util/rand/rand.go#L26-L31)

[the remark this answers is not in the document]
Comment (sbueringer) [reply cmd/clusterctl/main.go:29]: I'll drop it, there seems to be no need for it.
Thx, good finding :)

File: cmd/clusterctl/client/cluster/crd_migration.go (hunk 1 above)
Comment (sbueringer) [reply cmd/clusterctl/client/cluster/crd_migration.go:163]: Looks like no changes to managed fields

**Comment by fabriziopandini** [comment]
/lgtm
/approve

/hold 
for @MaxRink @schrej @furkatgofurov7  to make another round of tests

**Comment by furkatgofurov7** [comment]
> @MaxRink @schrej @furkatgofurov7 Latest version should work now. I verified it by upgrading CAPI: v0.3.14=>v0.4.8=>main (and dropped v1alpha3 from main).
> 
> Would be great if you can verify.



> /lgtm /approve
> 
> /hold for @MaxRink @schrej @furkatgofurov7 to make another round of tests

@sbueringer @fabriziopandini awesome work, thanks! I have tested it on Metal3 provider and it works pretty well :tada: , 

Here the results I got for reference:

**Before this patch:** Using the clusterctl binary built from the main branch and upgrading from v1a3==>v1a4=>v1beta1 (provider versions v1a4==>v1a5=>v1beta1):

```
./clusterctl upgrade apply --contract v1beta1
Checking cert-manager version...
Deleting cert-manager Version="v1.5.3"
Installing cert-manager Version="v1.8.2"
I0629 12:16:03.291677 3755760 request.go:601] Waited for 1.032853447s due to client-side throttling, not priority and fairness, request: GET:https://127.0.0.1:38293/apis/bootstrap.cluster.x-k8s.io/v1alpha4?timeout=30s
Waiting for cert-manager to be available...
Performing upgrade...
Scaling down Provider="infrastructure-metal3" Version="v0.5.10" Namespace="capm3-system"
Deleting Provider="infrastructure-metal3" Version="v0.5.10" Namespace="capm3-system"
Installing Provider="infrastructure-metal3" Version="v1.1.2" TargetNamespace="capm3-system"
Error: action failed after 10 attempts: failed to patch provider object: CustomResourceDefinition.apiextensions.k8s.io "metal3clusters.infrastructure.cluster.x-k8s.io" is invalid: status.storedVersions[0]: Invalid value: "v1alpha4": must appear in spec.versions
```


**After this patch:** next, with the clusterctl binary built using the changes on this branch:

```
./clusterctl upgrade apply --contract v1beta1
Checking cert-manager version...
Cert-manager is already up to date
Performing upgrade...
CR migration required kind="Metal3Cluster" storedVersionsToDelete="v1alpha4" storedVersionsToPreserve="v1alpha5"
Migrating CRs, this operation may take a while... kind="Metal3Cluster"
CR migration required kind="Metal3DataClaim" storedVersionsToDelete="v1alpha4" storedVersionsToPreserve="v1alpha5"
Migrating CRs, this operation may take a while... kind="Metal3DataClaim"
CR migration required kind="Metal3Data" storedVersionsToDelete="v1alpha4" storedVersionsToPreserve="v1alpha5"
Migrating CRs, this operation may take a while... kind="Metal3Data"
CR migration required kind="Metal3DataTemplate" storedVersionsToDelete="v1alpha4" storedVersionsToPreserve="v1alpha5"
Migrating CRs, this operation may take a while... kind="Metal3DataTemplate"
CR migration required kind="Metal3Machine" storedVersionsToDelete="v1alpha4" storedVersionsToPreserve="v1alpha5"
Migrating CRs, this operation may take a while... kind="Metal3Machine"
CR migration required kind="Metal3MachineTemplate" storedVersionsToDelete="v1alpha4" storedVersionsToPreserve="v1alpha5"
Migrating CRs, this operation may take a while... kind="Metal3MachineTemplate"
CR migration required kind="Metal3Remediation" storedVersionsToDelete="v1alpha4" storedVersionsToPreserve="v1alpha5"
Migrating CRs, this operation may take a while... kind="Metal3Remediation"
CR migration required kind="Metal3RemediationTemplate" storedVersionsToDelete="v1alpha4" storedVersionsToPreserve="v1alpha5"
Migrating CRs, this operation may take a while... kind="Metal3RemediationTemplate"
Scaling down Provider="infrastructure-metal3" Version="v0.5.10" Namespace="capm3-system"
Deleting Provider="infrastructure-metal3" Version="v0.5.10" Namespace="capm3-system"
Installing Provider="infrastructure-metal3" Version="v1.1.2" TargetNamespace="capm3-system"
```
```

./clusterctl upgrade plan
Checking cert-manager version...
Cert-Manager is already up to date

Checking new release availability...

Latest release available for the v1beta1 API Version of Cluster API (contract):

NAME                    NAMESPACE                           TYPE                     CURRENT VERSION   NEXT VERSION
bootstrap-kubeadm       capi-kubeadm-bootstrap-system       BootstrapProvider        v1.1.4            Already up to date
control-plane-kubeadm   capi-kubeadm-control-plane-system   ControlPlaneProvider     v1.1.4            Already up to date
cluster-api             capi-system                         CoreProvider             v1.1.4            Already up to date
infrastructure-metal3   capm3-system                        InfrastructureProvider   v1.1.2            Already up to date

You are already up to date!
```

File: cmd/clusterctl/client/cluster/crd_migration.go [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+	}
+
+	// Otherwise a version that has been used as storage version will be dropped, so it is necessary to migrate all the
+	// objects and drop the storage version from the current CRD status before installing the new CRD.
+	// Ref https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#writing-reading-and-updating-versioned-customresourcedefinition-objects
+	// Note: We are simply migrating all CR objects independent of the version in which they are actually stored in etcd.
+	// This way we can make sure that all CR objects are now stored in the current storage version.
+	// Alternatively, we would have to figure out which objects are stored in which version but this information is not
+	// exposed by the apiserver.
+	storedVersionsToDelete := currentStatusStoredVersions.Difference(newVersions)
```
Comment (chrischdi) [inline cmd/clusterctl/client/cluster/crd_migration.go:124]: storedVersionsToDelete is only for information purposes.

Ok for me, just want to make sure we did not initially want to skip if there are none to delete.

Comment (sbueringer) [reply cmd/clusterctl/client/cluster/crd_migration.go:124]: I think we already do, if there are no versions to delete storedVersionsToDelete will be empty which is the same as:
```go
	if newVersions.HasAll(currentStatusStoredVersions.List()...) {
		log.V(2).Info("CRD migration check passed", "name", newCRD.Name)
		return false, nil
	}
```

Comment (chrischdi) [reply cmd/clusterctl/client/cluster/crd_migration.go:124]: Ack, so in that case we will never even reach this line 👍 

As mentioned: ok for me to keep it for the users to see what was removed :-)

**Review by ykakarap — COMMENTED** [review]
/lgtm

**Comment by sbueringer** [comment]
/hold cancel

Give the positive feedback + it would be nice to have it running in CI a bit before the v1.1.5 release

**Pull request merged**